---
title: "03_iranges"
author: "JR"
date: "2024-11-07"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Goal: to understand the basic usage of Iranges. These principles are core to almost
all analyses we will perform so it's worth taking sometime to review the fundamentals.

Our research goal is to see where peaks reside in the genome and how they change
across time points. So we need to learn how to keep track of where a peak is and 
learn how to find overlaps between start and stop intervals of peaks.

# Iranges allows one to index "windows" with properties. Say you had a vector
of all genes and wanted to track start and stop. Iranges can keep the start, 
stop and width of values. So this list of ranges (start and stop) for each 
gene-length is essentially a vector or each column a gene id. With this indexing
we can now apply several attributes to these ranges. For example if an enhancer 
or other feature resided in the gene we could also track the start, stop of that
"interval". 

In short, we can create a map of where each peak of chromatin accessibility resides
realtive to any other genomic feature of our choosing - we just have to keep track
of all this in the form of intervals (start, end). Then we can do a lot of analysis !

# Granges is even more specially aadapted for features in a given genome. All
the start, stops for each feature is given similar to Iranges. However, we could
have a big problem since we have 23 chromosomes in the genome! So simple start
and stop indexing isn't enough as there could be up to 23 redundant "intervals" 
that are actually unique. So we need to consider chromosome and coordinates. 
Worse we also should think about the strand. So GRanges works similar to Iranges
just having the advantage of indexing chromosome and strand to the start and stop.

# Now let's start with the basics using IRanges

```{r IRanges and GRanges packages}
library(IRanges)
? IRanges

library(GenomicRanges)
? GRanges

# These two packages both keep track and can be operated on are: 
# start
# end
# width
```


# Let's make a list of ranges and see what this looks like.
Think of each range as a "peak" start and stop. And each peak
is a peice of the genome that is accessible or open (to Tn5 transposon) !

```{r our first range}

# Let's use an example of a ATAC-peak-files each with three peaks. We can use
# this simple example to construct and intersect ranges -- a common procedure

ir1 <- IRanges(start = c(1,200, 1000), end = c(30, 299, 1200))

# take a look
ir1

# we can use a lot of Iranges functions to look at what we just made
start(ir1)
end(ir1)
width(ir1)

# Cool let's make a second set of peak ranges. We can think of it 
# as a replicate ATACseq with 3 peaks in slightly different position.

ir2 <- IRanges(start = c(17,100, 1100), end = c(49, 199, 1201))

# take a look
ir2 

# nice, so let's concatenate these into one range
ir3 <- c(ir1, ir2)

# take a look -- stacked on top
ir3

# The nice thing about IRanges as we can track values associated with each range
# for example we may want to name these replicate 1 and replicate 2
# let's give it a try

names(ir1) <- paste("replicate_1", "peak", 1:3, sep = " ")
ir1

# Nice we now have a new "level" to index that is the name of the range
# or @NAMES in environment. 


# Let's do the same for ir2.
names(ir2) <- paste("replicate_2", 1:3, sep = " ")
ir2
# now concatenate  
ir3 <- c(ir1, ir2)

# let's see
ir3

# nice we have our 'data set' nicely labeled and organized.

```
If you want to plot these you can look here for good code
Not really essential but good thought practice and plotting practice.
https://stackoverflow.com/questions/21506724/how-to-plot-overlapping-ranges-with-ggplot2


# finding overlaps in iranges
```{R overlaps in Iranges}
# first let's use union to "merge" peaks.
# Remaking range 1 since we changed above

ir1 <- IRanges(start = c(1,200, 1000), end = c(30, 299, 1200))
ir2 <- IRanges(start = c(17,100, 1100), end = c(49, 199, 1201))

# union (Iranges function) 
# note: forces the union even if not overlapping!
ov1 <- union(ir1, ir2)
ir1
ir2
# example: 1 is "union" with 17 and is 49 long
# example: 200 and 100 are forced to overlap and length of 200 etc

# let's look with ir3 to reference
ir3
ov1
# not exactly what we want but useful let's use intersect:
```

Let's use intersect which requires an overlap to make the intersection. 
```{R overlaps in Iranges}

# intersect (min overlap)
# will only result in overlapping ranges
ov3 <- intersect(ir1, ir2)

# let's look
ov3
ir3

# cool, so this is typically what we want for the min overlap
# example 1-30 overlaps with 17-49 = 17-30 is intersect (14 long)

# we can also look a this with countOveralps to get more info
# note that the counts of overlaps are recorded but not where!
countOverlaps(ir1, ir2)
ir3

# order matters -- but can be figured out with just one countOverlaps
countOverlaps(ir2, ir1)

```

Iranges is great if there is only one chromosome in the genome :) 
However, there are multiple chromosomes in the genome and there are two strands
of DNA that we need to keep track of.

This is a good primer on how intervals are considred and can be overlapped --
but now it's time to go back to our data and use GENOMIC RANGES !