---
title: "01_import_peaks"
author: "JR"
date: "2024-11-16"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(stringr)
library(GenomicRanges)
```

# GOAL: To learn how to import peak files as GRanges and find overlaps between samples!
# For review see previous lesson: 08_ATACseq_pipeline/04_granges.Rmd
We now have peak files for each individual time point. Let's learn how to import these files
and analyze them for overlaps between samples.

# Let's go Find your peak output files in: your_ATACseq_pipeline_output/bwa/merged_library>
Let's start with a simple question: what is the overlap between WT_0 and KO_0?
To do this we need rtarcklayer to import peak files as GRanges so we can find overlaps.

# First let's load in the the WT_0 and KO_0 peak files from MACS2 
```{r}
# reading in WT_0 ATACseq peak calls !

WT_0_Peaks <- rtracklayer::import("../08_ATACseq_pipeline/00_run_pipeline/00_pipeline_run/pipeline_run/bwa/merged_library/macs2/broad_peak/ESC_WT_control_0_REP1.mLb.clN_peaks.broadPeak")

# let's take a look at the data 
view(WT_0_Peaks)

# Now KO no dox (0 timepoint)
KO_0_Peaks <- rtracklayer::import("../08_ATACseq_pipeline/00_run_pipeline/00_pipeline_run/pipeline_run/bwa/merged_library/macs2/broad_peak/ESC_KO_control_0_REP1.mLb.clN_peaks.broadPeak")

# Let's take a look at what information we have !
view(KO_0_Peaks)
view(WT_0_Peaks)

# take a look at column names
head(WT_0_Peaks)
head(KO_0_Peaks)

# Cool so we see a lot of data is present
# IMPORTANT - note the pValue and qValue are the negative log10 of the pavlue
#  -log10(p-value), the values are always positive numbers
# For example, a value of 5 corresponds to an actual p-value of 10−510^{-5}10−5 - very significant
# Significance starts around .893 (which is 0.05)

```
# IMPORTANT: our peak files also contain non chromosomal contig regions of the genome !!
# IMPORTANT: to filter out non-cannonical chromsomes.

# Let's filter to cannonical chromosomes 
```{r}
#Let's take a look at one file to start (KO_0_Peaks):
seqnames(KO_0_Peaks)
# Lots of GL contigs we don't want, but we want to grab everything with a "chr" at start!

# Let's use grep to grab the enteries that start with "chr" indicated by "^chr"
KO_0_chr_peaks <- KO_0_Peaks[grepl("^chr", as.character(seqnames(KO_0_Peaks)))]
WT_0_chr_peaks <- WT_0_Peaks[grepl("^chr", as.character(seqnames(WT_0_Peaks)))]

# Let's take a look:
seqnames(KO_0_chr_peaks)
# We see now our seqnames values all start with chr
# don't be scared that the levels are still there for non chr contigs
seqlevels(KO_0_chr_peaks)

```


# Now let's explore how to overlap these files using findOverlaps!
```{r findOverlpas}
# Let's find overlaps using find overlaps function from genomic ranges GRanges
# This will tell us how many peaks are the same in both 0 time point ATACseq experiments.
ovf <- findOverlaps(WT_0_Peaks, KO_0_Peaks)
# Looking at stats in console
length(ovf)

# Let's do the same for our chromosome only files:
ovf_chr <- findOverlaps(WT_0_chr_peaks, KO_0_chr_peaks)

# Let's compare
length(ovf)
length(ovf_chr)

# interesting there are 54,170 with contigs and 54,064 with just chr
# pretty close but GOOD TO NOTE !
# We will move forward with the chromosome filtered versions for class !
```

# Now let's compare the query file (first @from) to the subject file (second @to) 

```{r findOverlpas}
# Let's look at how many unique peaks there were in KO_0_Peaks (@to)
length(unique(ovf_chr@to))
# So 53,747 peaks in KO_0_Peaks were unique comprising the 54,064 overlaps.

# now let's look at WT_0_Peaks
length(unique(ovf_chr@from))
# 53,827 peaks made up the 54,064 overlaps.


# Let's look at how many had multiple overlaps
table(table(ovf_chr@from))
# same for KO_0
table(table(ovf_chr@to))

# Ok numbers add up and not unusual (like crazy peaks - remember them)
# Meaning it's about the same number of overlaps both ways with not to many multiple overlaps
```
# Overall not bad overlap between samples !
Now let's learn a skill to import a bunch of peak files (all)
as Genomic Ranges (GRanges)! We can do that with a couple handy 
new tools for our kit.

# Let's import all peak files !

```{r base file path to peaks}

# First let's set a file path to where all our peak files are.
# This is very helpful in the long run to not have huge filepaths everywhere :)
peak_path <- "/scratch/Shares/rinnclass/MASTER_CLASS/lessons/08_ATACseq_pipeline/00_run_pipeline/00_pipeline_run/pipeline_run/bwa/merged_library/macs2/broad_peak"

# Now let's make a list of filepaths to each peak file -- "list.files"
# we add a pattern parameter to just load .broadPeaks as there are many other types.

fl <- list.files(peak_path, 
                 full.names=TRUE, pattern = ".broadPeak")

# let's see 
fl

# This gives us an index 1:n where each peak file path is a character vector.
# let's try this to grab just one protein (can be the one from your group)
fl_KO <- fl[grep("KO_control", fl)]

# cool now we can subset this to any sample of our choosing - and directly connect file path
pattern <- "(KO|WT)_control_\\d+"

```

# extracting sample name from file name from all files !

```{r extract sample name from file}

# First set up a simple function to get warmed up !
# We will go through each 
sample_name <- sapply(fl, function(y){
 y <-  str_extract(y, "(KO|WT)_control_\\d+")
  
})

sample_name
# Nice so we can see a start of a function that grabs name from file path

```

# Let's make an import peaks function since this is a common/repetitive task
```{r making import peaks function}

# let's start with what we did above and create a custom function called "import_peaks"

import_peaks <- function(consensus_file_path) {
  
  # List all MACS2 broadPeak files in the directory
  peak_files <- list.files(consensus_file_path, full.names = TRUE, pattern = ".broadPeak")
  
  # Extract sample names directly from file names
  sample_names <- sapply(peak_files, function(file) {
    str_extract(file, "(KO|WT)_control_\\d+")
  })
  
  # Initialize an empty list to store filtered GRanges objects
  peak_list <- list()
  
  # Import, filter, and store GRanges objects
  for (i in seq_along(peak_files)) {
    # Import peaks as GRanges
    peaks <- rtracklayer::import(peak_files[i])
    
    # Filter to keep only seqnames starting with "chr"
    filtered_peaks <- peaks[grepl("^chr", as.character(seqnames(peaks)))]
    
    # Add filtered peaks to the list with a sample-specific name
    peak_list[[sample_names[i]]] <- filtered_peaks
  }
  
  return(peak_list)
}
```


# Let's test run our function !
```{r running import peaks function}

# Let's run our new custom function import_peaks
my_peaks <- import_peaks(consensus_file_path = peak_path)

# NICE we just imported all the peak files and names them - take a look!
# Each row is a GRange object so this is a list of GRanges - so much packed in !

# Le't see how many peaks each file has:
num_peaks <- sapply(my_peaks, length) %>% as.data.frame()

# Double checking we only have cannonical chromosomes by indexing one of the GRanges in the list
seqnames(my_peaks$KO_control_0)

# Nice good to go !
```
# Nice you now have a function called import_peaks! 
You can point it to any directory with peak files and it 
will load them all right up as GRanges and ready to do overlaps.


*********************
EXCERCISE
*********************

Create a function called import_peaks in your class_functions.R file
So it can be sourced next class with source(.R Function)
We previously called it 00_our_first_function_lesson.R 
But now maybe a good time to rename it :) 















