---
title: "04_differential_peaks"
author: "JR"
date: "2024-11-28"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(GenomicRanges)
source("../../util/00_our_first_function_lesson.R")
```

# Goal: To find ATACseq peaks that change upon exposure to dox.
First we will attempt a simple overlap of 0 time point (no dox)
to the rest of the time points. This will tell us which peaks existed 
at time 0 and dissapeared and those that didn't exist at time 0 and appeared!


# Loading environment from 03: peak/overlap objects, genome_feature objects
```{r loading objects from 03_peaks_to_genome_feature.rmd}

# Loading our prior work !
load("results/key_mouse_atac_data.Rdata")

# Nice - good to go !
```

# Find peaks that are common to 0 time points
```{r}

# Setting object to 0 time point
non_dox_samples <- my_peaks[c("KO_control_0", "WT_control_0")]

# find common peaks
non_dox_common_peaks <- find_common_peaks(non_dox_samples)
length(non_dox_common_peaks)
# random file test
length(my_peaks$KO_control_0)

# Nice we have the overlaps of our non-dox, let's do the same for dox !
# first grep out non 0 time point names
dox_samples <- names(my_peaks)[!grepl("_0$", names(my_peaks))]

# Now indexing to make a list of GRanges of non-0 (no dox) time points
dox_peaks <- my_peaks[dox_samples]

# finding common overlaps
dox_common_peaks <- find_common_peaks(dox_peaks)
length(dox_common_peaks)
# random single example
length(my_peaks$KO_control_30)

# Nice we are now set to compare non-dox vs dox !
```

# Now let's compare 0 to all other time points for peaks that do and do not overlap 
```{r comparing 0 peaks to all other time points for overlaps}

# create input list to find_common_peaks
dox_compare_list <- list(non_dox = non_dox_common_peaks, dox = dox_common_peaks)
# run find_common_peaks
dox_non_dox_ov <- find_common_peaks(dox_compare_list)
length(dox_common_peaks)
length(dox_non_dox_ov)
# This means that 37435 of 38984 peaks overlapped
# Percent overlap of dox in non-dox peaks
37435/38984
# 96% WOW! that means almost all peaks in dox are in non-dox
# how many unique dox peaks not in present before dox
37435 -38984
# 1549 unique peaks

# Checking the reverse comparison of non-dox peaks in dox
length(non_dox_common_peaks)
length(dox_non_dox_ov)
# Now we see 37435 of 53747 overlap
# percentage non-dox in dox :
37435/53747
# number peaks in non-dox and not in dox
37435 - 53747
# WOW 16312 peaks are dissapear in dox treatment
# This is either really cool or suspicious :) 

```
# Result: almost all peaks common in dox were also present in non-dox
# Result: 37435 dox common peaks of the 38984 non-dox peaks overlapped (96%)
# Result: 16312 peaks found in non-dox that were not found in dox common
# Result: 1549 unique peaks in dox and not present in non-dox
# Result: Overall bias to more non-dox specific peaks (could it be due to sample num?)

# Hypothesis: There is a Massive chromatin closing upon dox exposure?
A large number of chromatin accessibility sites (ATAC peaks) sites
are "closing" upon dox exposure. WE find 16K sites that close from 0
across the rest of the time points. We also observe 1.5K peaks unique to 
dox exposure and not present before.

# Next step: let's see : prepare our new results for viewing in IGV
First we need to make .bed files of peaks only in dox (1.5K)
or only in non-dox(16K).

# Step 1: create quick function to grab unique peaks to dox or non-dox
Here we create a small but handy function to select any
peaks we want from any comparison.
```{r create find my peaks function for dox and non-dox specific peaks}

# create find_my_peaks function
find_my_peaks <- function(common_peaks, original_peaks) {
  # Find overlaps
  overlaps <- findOverlaps(original_peaks, common_peaks)
  
  # Identify peaks in original_peaks that are not in common_peaks
  unique_peaks <- original_peaks[-queryHits(overlaps)]
  
  # return peaks in condition selected
  return(unique_peaks)
}


# Peaks unique to non_dox
unique_to_non_dox <- find_my_peaks(dox_non_dox_ov, non_dox_common_peaks)
length(unique_to_non_dox)
# Exact same number we had before!

# Peaks unique to dox
unique_to_dox <- find_my_peaks(dox_non_dox_ov, dox_common_peaks)
length(unique_to_dox)
# A little longer but super close - we can tell in IGV soon


```

# Now that we have dox and non-dox peaks let's overlap with RNAseq hits!
First we will need to load our sig genes and then intersect
The data we need from RNAseq is in 06_Differential_expression_analyses/results/time_point_res_df.Rdata
```{r overlap of non-dox peaks and sig RNAseq changes}

load("/scratch/Shares/rinnclass/MASTER_CLASS/lessons/06_Differential_expression_analyses/results/counts_results_DESEQ/DESEQ_results.rdata")

# make gr_list input for find_common_peaks custom function
gr_list_gene_promoter_non_dox_ov <- list( gene_promoters = gene_promoters, non_dox_promoters = unique_to_non_dox)

# find overlaps
non_dox_gene_promoter_ov <- find_common_peaks(gr_list_gene_promoter_non_dox_ov)
# 2178 peaks overlap gene promoters out of 16,312

# Now find same gene_id in RNAseq resuts
sig_rnaseq_atac_non_dox <- non_dox_gene_promoter_ov[non_dox_gene_promoter_ov$gene_id %in% filtered_res_df$gene_id]

# how many genes overlap between RNAseq and ATACseq non-dox peaks!
length(sig_rnaseq_atac_non_dox)
#47 not many ... let's look (see below)

```

# Guess what we do next?
Look at the peaks in the raw data (IGV) :) !
Let's look at peaks called as different in dox and non-dox in IGV
```{r exporting to bed files to view in IGV}

# export non-dox peaks
rtracklayer::export(unique_to_non_dox, con = "results/unique_to_non_dox.bed", format = "BED")

# export dox peaks
rtracklayer::export(unique_to_dox, con = "results/unique_to_dox.bed", format = "BED")

# Let's export the sig atac - rna seq overlap
df_sig_rnaseq_atac_non_dox <- as.data.frame(sig_rnaseq_atac_non_dox)
write_csv(df_sig_rnaseq_atac_non_dox,"results/rnaseq_atacseq_overlaps.csv")

```


############################
Exercise: View peaks in IGV
############################

# Now let's explore our results in IGV asking the following:

(a) Do the peaks called unique to dox look bigger in non-dox than dox?

(b) Look at a few examples of ATACseq/RNAseq overlap gene names in IGV
- do the peaks look compeling? Do they go in the same direction as gene?


