---
title: "05_R_genome_GRanges"
author: "JR"
date: "2024-11-13"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = F)
library(GenomicRanges)
# library(Gviz)
library(IRanges)

```

# GOAL: Get to know the features of your genome !
The genome is a 3 Billion letter sequence that was a heroic effort
accomplished in the year 2000. Since the goal has been to "annotate"
features of the human genome (such as long noncoding RNA my favorite :)

It all follows the same logic we learned in GRanges - just chromosome
start and stop tells us where a feature is. In fact there are entire consortiums to annotate the human genome. 

These feature to position maps are compiled in the GTF file we used
as input into both the ATACseq and RNAseq NF_CORE pipelines. 

# Let's get to know our genome in the form of GRanges !
https://uswest.ensembl.org/info/website/upload/gff.html

# (1) We Need GEncode GTF of annotations for this class.  Let's go ge features of the human genome ! I have downloaded this from gencode and stored for everyone to use the same annotation file.

# IMPORTANT: Genome annotations change.
So you always want to keep track of what version you are using 
otherwise your data alignments could be megabases away in differnet
genomic annotaitons. 

If not in class please download Gencode annotations gtf here : https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_32/


## Let's get the GENCODE : VERSION GTF of MOUSE Genome Annotations
# IMPORTANT: we are using Mouse Genome version 25
```{r loading in GENCODE GRanges annotation file}

# we use rtracklayer::import to turn the annotations into start and stops (Iranges) w/ chr (Granges)This takes ~5min using GRanges Package.

gencode_gr <- rtracklayer::import("/scratch/Shares/rinnclass/MASTER_CLASS/GENOMES/M25/gencode.vM25.annotation.gtf")

# let's take a look @metadata to see what we have in here

```

Now let's look at the gene_id indexable part of this GRanges
# seqnames
```{r seqnames}

?seqnames
  
seqnames(gencode_gr) %>% 
  head()

# we can see that "levels" are the chromosome names and are a meta-level

# Note ChrM !
```

let's get more info with table to summary using tidy verse %>%
# summary / table a GRanges object
```{R summary/table GRanges}

table(seqnames(gencode_gr)) %>%
  head()
# This is telling us how many features are associated with each chromosome.

table(seqnames(gencode_gr)) %>% 
  summary()
# this tells us how many features there are in the genome.

# Let's find out how many features the mitochondrial chr has (22th level - mouse (20) has less chromosomes than humans (23)
table(seqnames(gencode_gr))[22] 

# we see this is the mitochondrial chromosome and has 153 elements

```

# Indexing Granges (@elementMetadata$INDEXING)
```{R indexing GRanges}

# we see a lot of good infromation in @elementMetadata -- let's index it.

# let's see how many 'gene_id' values there are by indexing
table(gencode_gr@elementMetadata$gene_id) %>% 
  head()
# This tell us how many features are associated with each gene id.

# NOTE gene_id is ensemble gene_id annotation. The (.) at the end of the number indicates which isoform of a gene.

# Let's see what we get with gene_name
table(gencode_gr@elementMetadata$gene_name) %>% 
  head(100)

# Typical gene symbols 
table(gencode_gr@elementMetadata$transcript_name) %>%
  head()

# transcript_id

# it's basically memorization of what is a gene name or gene id 

# if you wanted to look at this more expanisvely (instead of just head lines)
# you can make it into an object

gencode_info <- table(gencode_gr@elementMetadata$gene_name) %>% 
 as.data.frame()

# cool now you can look for any feature for your favorite genes
```


Let's analyze some of the features of our genome!
# Making a data frame of items in GRanges object
```{r GRanges indexing to data.frame}

# above we tested different indexes and made a dataframe of one feature
# now let's do this multiple times at once!

gencode_gr_length <- data.frame("gene_id" = gencode_gr@elementMetadata$gene_id,
                           "gene_name" = gencode_gr@elementMetadata$gene_name,
                           "gene_type" = gencode_gr@elementMetadata$gene_type,
                           "type" = gencode_gr@elementMetadata$type,
                           "start" = gencode_gr@ranges@start,
                           "width" = gencode_gr@ranges@width,
                           "chr" = gencode_gr@seqnames)
                           
```

# Let's save our gencode_gr and gencode_gr_length objects for the next lesson - and good book keeping!

```{r save our GRanges}

# The key objects we need moving forward (to save that 5min rtracklayer::import) are : 
# Gencode_gr : has all the genome features as GRanges object
# Gencode_gr_length: is in more of a data frame format with properties selected above
# -- feel free to pick more in the future ;)

write_rds(gencode_gr, "results/mv25_gencode_GRanges.rds")
write_rds(gencode_gr_length, "results/mv25_gencode_gr_length.rds")
save(gencode_gr, gencode_gr_length, file = "results/mv25_gencode_GRanges.Rdata")

```



####################
Excercise
###################

# Cool now we have a lot of info about genome features in a data frame!
Now go explore these questions below:

1) What is the longest gene?
2) What is the shortest?
3) How long is the Firre Gene? 
4) What is the lenght of the longest Firre exon?
5) What is the largest exon in the mouse genome?
6) What features or annotations have a width of 1?

